group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+    }
+
+    /**
+     * This class maintains some data structures to simplify lookup of partition movements among consumers.
+     * At each point of time during a partition rebalance it keeps track of partition movements corresponding to each topic,
+     * and also possible movement (in form a <code>MemberPair</code> object) for each partition.
+     */
+    private static class PartitionMovements {
+        private final Map<Uuid, Map<MemberPair, Set<TopicIdPartition>>> partitionMovementsByTopic = new HashMap<>();
+        private final Map<TopicIdPartition, MemberPair> partitionMovementsByPartition = new HashMap<>();
+
+        private MemberPair removeMovementRecordOfPartition(TopicIdPartition partition) {
+            MemberPair pair = partitionMovementsByPartition.remove(partition);
+
+            Uuid topic = partition.topicId();
+            Map<MemberPair, Set<TopicIdPartition>> partitionMovementsForThisTopic = partitionMovementsByTopic.get(topic);
+            partitionMovementsForThisTopic.get(pair).remove(partition);
+            if (partitionMovementsForThisTopic.get(pair).isEmpty())
+                partitionMovementsForThisTopic.remove(pair);
+            if (partitionMovementsByTopic.get(topic).isEmpty())
+                partitionMovementsByTopic.remove(topic);
+
+            return pair;
+        }
+
+        private void addPartitionMovementRecord(TopicIdPartition partition, MemberPair pair) {
+            partitionMovementsByPartition.put(partition, pair);
+
+            Uuid topic = partition.topicId();
+            if (!partitionMovementsByTopic.containsKey(topic))
+                partitionMovementsByTopic.put(topic, new HashMap<>());
+
+            Map<MemberPair, Set<TopicIdPartition>> partitionMovementsForThisTopic = partitionMovementsByTopic.get(topic);
+            if (!partitionMovementsForThisTopic.containsKey(pair))
+                partitionMovementsForThisTopic.put(pair, new HashSet<>());
+
+            partitionMovementsForThisTopic.get(pair).add(partition);
+        }
+
+        private void movePartition(TopicIdPartition partition, String oldOwner, String newOwner) {
+            MemberPair pair = new MemberPair(oldOwner, newOwner);
+
+            if (partitionMovementsByPartition.containsKey(partition)) {
+                // This partition was previously moved.
+                MemberPair existingPair = removeMovementRecordOfPartition(partition);
+                assert existingPair.dstMemberId.equals(oldOwner);
+                if (!existingPair.srcMemberId.equals(newOwner)) {
+                    // The partition is not moving back to its previous consumer.
+                    addPartitionMovementRecord(partition, new MemberPair(existingPair.srcMemberId, newOwner));
+                }
+            } else
+                addPartitionMovementRecord(partition, pair);
+        }
+
+        private TopicIdPartition getTheActualPartitionToBeMoved(TopicIdPartition partition, String oldOwner, String newOwner) {
```

Comment:
I have a hard time to really understand what this method does. Could you help me out?

Reply:
yep sure, this method is determining which partition should actually be moved to a new owner, considering the history of partition movements and ownership changes. If a partition has been moved before or if there's a recorded reverse movement, it takes these into account to decide the next move. I can explain better offline if required.

Reply:
done thanks!